state-reconstruct-fetcher/src/l1_fetcher.rs

```diff
@@ -243,81 +233,120 @@ impl L1Fetcher {
+                                            if cancellation_token.is_cancelled() {
+                                                tracing::debug!("Shutting down tx sender...");
+                                            } else {
+                                                tracing::error!("Cannot send tx hash: {e}");
+                                                cancellation_token.cancel();
+                                            }
+
+                                            return current_l1_block_number.as_u64();
+                                        }
+
+                                        previous_hash = Some(tx_hash);
                                     }
 
-                                    return current_l1_block_number.as_u64();
+                                    latest_l2_block_number = new_l2_block_number;
                                 }
+                            } else {
+                                tokio::time::sleep(Duration::from_secs(LONG_POLLING_INTERVAL_S))
+                                    .await;
+                                continue;
+                            };
+
+                            metrics.lock().await.latest_l1_block_num =
+                                current_l1_block_number.as_u64();
+
+                            let next_l1_block_number =
+                                current_l1_block_number + U64::from(BLOCK_STEP);
+                            if next_l1_block_number > end_block_number {
+                                if current_l1_block_number < end_block_number {
+                                    current_l1_block_number = end_block_number;
+                                } else {
+                                    if disable_polling {
+                                        tracing::debug!("Fetching finished...");
+                                        return current_l1_block_number.as_u64();
+                                    }
 
-                                previous_hash = Some(tx_hash);
+                                    current_l1_block_number = end_block_number + U64::one();
+                                }
+                            } else {
+                                current_l1_block_number = next_l1_block_number;
+                            }
+                        }
+                    } else if let Ok(new_end) = L1Fetcher::retry_call(
+                        || provider_clone.get_block(BlockNumber::Latest),
+                        L1FetchError::GetEnd,
+                    )
+                    .await
+                    {
+                        if let Some(found_block) = new_end {
+                            if let Some(end_block_number) = found_block.number {
+                                end_block = Some(end_block_number);
+                                metrics.lock().await.last_l1_block = end_block_number.as_u64();
```

Comment:
I admit I don't see the chained syntax here - it can't be just `Some(Some(end_block_number))`, because the `found_block` must be dereferenced - so I just wrote it simply... :-)

Reply:
I was more thinking of chaining the `Ok(Some(found_block))`, since at the the moment, `Some(found_block)` is just matched on `new_end` and thus redundant

Reply:
I don't think it's redundant: if the awaited `retry_call(get_block` succeeds, the code tries to update `end_block` (which might still fail but will be retried immediately), else the network error is logged and the program waits - those are different paths...